src/range_query/sqrt_decomp.rs

```diff
@@ -116,31 +118,67 @@ impl PiecewiseLinearFn {
     pub fn with_merge_threshold(merge_threshold: usize) -> Self {
         Self {
             sorted_lines: vec![],
+            intersections: vec![],
             recent_lines: vec![],
             merge_threshold,
         }
     }
 
     /// Replaces this function with the minimum of itself and a provided line
-    pub fn min_with(&mut self, slope: i64, intercept: i64) {
+    pub fn min_with(&mut self, slope: f64, intercept: f64) {
         self.recent_lines.push((slope, intercept));
     }
 
     fn update_envelope(&mut self) {
         self.recent_lines.extend(self.sorted_lines.drain(..));
-        self.recent_lines.sort_unstable();
-        for (slope, intercept) in self.recent_lines.drain(..) {
-            // TODO: do convex hull trick algorithm
-            self.sorted_lines.push((slope, intercept));
+        self.recent_lines
+            .sort_unstable_by(|x, y| y.partial_cmp(&x).unwrap());
+        self.intersections.clear();
+
+        for (m1, b1) in self.recent_lines.drain(..) {
+            while let Some(&(m2, b2)) = self.sorted_lines.last() {
+                // If slopes are equal, the later line will always have lower
+                // intercept, so we can get rid of the old one.
+                if (m1 - m2).abs() > 1e-10f64 {
```

Comment:
Is this the idiomatic way to test floats for equality? If so, we may want a library-global constant epsilon.

Reply:
Hmm I don't know. Even for integer inputs, if they're big enough, it might be hard to distinguish equal from unequal slopes, and it's not totally clear to me what effect the choice of epsilon will have. We can do the analysis later if we want, but this looks good enough for the PR.